File: Factory/AbstractFactory/TypeScript/src/Pizza/Pizza.ts
import Dough from '../Ingredients/Dough';
import Sauce from '../Ingredients/Sauce';
import Veggies from '../Ingredients/Veggies';
import Cheese from '../Ingredients/Cheese';
import Pepperoni from '../Ingredients/Pepperoni';
import Clams from '../Ingredients/Clams';

export default abstract class Pizza {
  name: string;

  dough: Dough;
  sauce: Sauce;
  veggies: Veggies[];
  cheese: Cheese;
  pepperoni: Pepperoni;
  clams: Clams;

  abstract prepare(): void;

  bake(): void {
    console.log('Bake for 25 minutes at 350');
  }

  cut(): void {
    console.log('Cutting the pizza into diagonal slices');
  }

  box(): void {
    console.log('Placing the pizza into an official PizzaStore box');
  }

  setName(name: string) {
    this.name = name;
  }

  getName(): string {
    return this.name;
  }

  toString(): string {
    return `${this.name}\n${this.dough ? `${this.dough}\n` : ''}${
      this.sauce ? `${this.sauce}\n` : ''
    }${this.cheese ? `${this.cheese}\n` : ''}${
      this.veggies ? `${this.veggies}\n` : ''
    }${this.clams ? `${this.clams}\n` : ''}${
      this.pepperoni ? `${this.pepperoni}\n` : ''
    }`;
  }
}

File: Factory/AbstractFactory/Java/src/Pizza/Pizza.java
package Pizza;

import Ingredients.*;

public abstract class Pizza {
    String name;

    Dough dough;
    Sauce sauce;
    Veggies veggies[];
    Cheese cheese;
    Pepperoni pepperoni;
    Clams clams;

    public abstract void prepare();

    public void bake(){
        System.out.println("Bake for 25 minutes at 350");
    }

    public void cut(){
        System.out.println("Cutting the pizza into diagonal slices");
    }

    public void box(){
        System.out.println("Placing the pizza in an official PizzaStore box");
    }

    public void setName(String name){
        this.name = name;
    }

    public String getName(){
        return this.name;
    }

    public String toString(){
        StringBuffer pizzaString = new StringBuffer();
        pizzaString.append(name + "\n");
        if(dough != null){
            pizzaString.append(dough);
            pizzaString.append("\n");
        }
        if(sauce != null){
            pizzaString.append(sauce);
            pizzaString.append("\n");
        }
        if(cheese != null){
            pizzaString.append(cheese);
            pizzaString.append("\n");
        }
        if(veggies != null){
            for(int i = 0; i< veggies.length; i++){
                pizzaString.append(veggies[i]);
                if(i < veggies.length - 1){
                    pizzaString.append(", ");
                }
            }
            pizzaString.append("\n");
        }
        if(clams != null){
            pizzaString.append(clams);
            pizzaString.append("\n");
        }
        if(pepperoni != null){
            pizzaString.append(pepperoni);
            pizzaString.append("\n");
        }
        return pizzaString.toString();
    }
}
